src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -263,14 +284,70 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
           // failure, it may have been an application-level error and it may
+          rpc::ErrorType error_type =
+              is_actor ? rpc::ErrorType::ACTOR_DIED : rpc::ErrorType::WORKER_DIED;
+          error_type = cancelled ? rpc::ErrorType::TASK_CANCELLED : error_type;
           // not make sense to retry the task.
-          task_finisher_->PendingTaskFailed(
-              task_id,
-              is_actor ? rpc::ErrorType::ACTOR_DIED : rpc::ErrorType::WORKER_DIED,
-              &status);
+          task_finisher_->PendingTaskFailed(task_id, error_type, &status);
         } else {
           task_finisher_->CompletePendingTask(task_id, reply, addr.ToProto());
         }
       }));
 }
+
+Status CoreWorkerDirectTaskSubmitter::KillTask(TaskSpecification task_spec,
+                                               bool force_kill, int num_tries) {
+  RAY_LOG(INFO) << "Killing task: " << task_spec.TaskId()
+                << ". Attempts left: " << num_tries;
+  const SchedulingKey scheduling_key(
+      task_spec.GetSchedulingClass(), task_spec.GetDependencies(),
+      task_spec.IsActorCreationTask() ? task_spec.ActorCreationId() : ActorID::Nil());
+  std::shared_ptr<rpc::CoreWorkerClientInterface> client = nullptr;
+  {
+    absl::MutexLock lock(&mu_);
+    auto scheduled_tasks = task_queues_.find(scheduling_key);
+    // See if task has not been shipped yet
+    if (scheduled_tasks != task_queues_.end()) {
+      for (auto spec = scheduled_tasks->second.begin();
+           spec != scheduled_tasks->second.end(); spec++) {
+        if (spec->TaskId() == task_spec.TaskId()) {
+          scheduled_tasks->second.erase(spec);
+
+          if (scheduled_tasks->second.empty()) {
+            task_queues_.erase(scheduling_key);
+          }
+          // Try to cancel lease requests
+          task_finisher_->PendingTaskFailed(task_spec.TaskId(),
+                                            rpc::ErrorType::TASK_CANCELLED);
+          return Status::OK();
+        }
+      }
+    }
+
+    cancelled_tasks_.emplace(task_spec.TaskId());
+    auto rpc_client = sent_tasks_.find(task_spec.TaskId());
+    // Checks for RPC handle for worker
+    if (rpc_client != sent_tasks_.end() &&
+        client_cache_.find(rpc_client->second) != client_cache_.end()) {
+      client = client_cache_.find(rpc_client->second)->second;
+    }
+  }
+
+  if (client == nullptr) {
+    return Status::OK();
+  }
+
+  auto request = rpc::KillTaskRequest();
+  request.set_intended_task_id(task_spec.TaskId().Binary());
+  request.set_force_kill(force_kill);
+  RAY_UNUSED(client->KillTask(
+      request, [this, task_spec, force_kill, num_tries](const Status &status,
+                                                        const rpc::KillTaskReply &reply) {
+        if (status.ok() && !reply.attempt_succeeded() && num_tries > 0) {
+          usleep(RayConfig::instance().cancellation_retry_ms() * 1000);
+          RAY_UNUSED(KillTask(task_spec, force_kill, num_tries - 1));
```

Comment:
Discussed offline, but I realized that this is actually not a good idea since the assigned worker could fail and the task could get resubmitted before we receive this reply. So actually this seems fine, but we should add a unit test for that case.